src/OVAL/probes/unix/linux/fwupdsecattr_probe.c

```diff
@@ -0,0 +1,361 @@
+
+	return ret;
+}
+
+/**
+ * fwupd_security_attr_result_to_string:
+ * @result: security attribute result, e.g. %FWUPD_SECURITY_ATTR_RESULT_ENABLED
+ *
+ * Returns the printable string for the result enum.
+ *
+ * Returns: string, or %NULL
+ *
+ * Since: 1.5.0
+ **/
+static const char *
+fwupd_security_attr_result_to_string(FwupdSecurityAttrResult result)
+{
+	if (result == FWUPD_SECURITY_ATTR_RESULT_VALID)
+		return "valid";
+	if (result == FWUPD_SECURITY_ATTR_RESULT_NOT_VALID)
+		return "not-valid";
+	if (result == FWUPD_SECURITY_ATTR_RESULT_ENABLED)
+		return "enabled";
+	if (result == FWUPD_SECURITY_ATTR_RESULT_NOT_ENABLED)
+		return "not-enabled";
+	if (result == FWUPD_SECURITY_ATTR_RESULT_LOCKED)
+		return "locked";
+	if (result == FWUPD_SECURITY_ATTR_RESULT_NOT_LOCKED)
+		return "not-locked";
+	if (result == FWUPD_SECURITY_ATTR_RESULT_ENCRYPTED)
+		return "encrypted";
+	if (result == FWUPD_SECURITY_ATTR_RESULT_NOT_ENCRYPTED)
+		return "not-encrypted";
+	if (result == FWUPD_SECURITY_ATTR_RESULT_TAINTED)
+		return "tainted";
+	if (result == FWUPD_SECURITY_ATTR_RESULT_NOT_TAINTED)
+		return "not-tainted";
+	if (result == FWUPD_SECURITY_ATTR_RESULT_FOUND)
+		return "found";
+	if (result == FWUPD_SECURITY_ATTR_RESULT_NOT_FOUND)
+		return "not-found";
+	if (result == FWUPD_SECURITY_ATTR_RESULT_SUPPORTED)
+		return "supported";
+	if (result == FWUPD_SECURITY_ATTR_RESULT_NOT_SUPPORTED)
+		return "not-supported";
+	return NULL;
+}
+
+int fwupdsecattr_probe_main(probe_ctx *ctx, void *arg)
+{
+	SEXP_t *val, *item, *ent, *probe_in;
+	oval_schema_version_t over;
+	char *stream_id = NULL;
+	const char *hsi_result_str;
+	uint64_t hsi_result = UINT64_MAX;
+
+	// arg is NULL if regex compilation failed
+	if (arg == NULL) {
+		return PROBE_EINIT;
+	}
+
+	probe_in = probe_ctx_getobject(ctx);
+	if (probe_in == NULL)
+		return PROBE_ENOOBJ;
+
+
+	over = probe_obj_get_platform_schema_version(probe_in);
+
+	ent = probe_obj_getent (probe_in, "stream-id", 1);
+	if (ent == NULL) {
+		return (PROBE_ENOENT);
+	}
+
+	val = probe_ent_getval (ent);
+	if (val == NULL) {
+		dD("%s: no value", "name");
+		SEXP_free (ent);
+		return (PROBE_ENOVAL);
+	}
+
+	stream_id = SEXP_string_cstr (val);
+	SEXP_free (val);
+	SEXP_free(ent);
+
+	DBusError dbus_error;
+	DBusConnection *dbus_conn;
+
+	if (LIST_EMPTY(&hsi_result_cache)) {
+
+		dbus_error_init(&dbus_error);
+		dbus_conn = connect_dbus();
+
+		if (dbus_conn == NULL) {
+			dbus_error_free(&dbus_error);
+			SEXP_t *msg = probe_msg_creat(OVAL_MESSAGE_LEVEL_INFO, "DBus connection failed, could not identify systemd units.");
+			probe_cobj_set_flag(probe_ctx_getresult(ctx), ctx->offline_mode == PROBE_OFFLINE_NONE ? SYSCHAR_FLAG_ERROR : SYSCHAR_FLAG_NOT_COLLECTED);
+			probe_cobj_add_msg(probe_ctx_getresult(ctx), msg);
+			SEXP_free(msg);
+			return 0;
+		}
+
+		get_all_security_attributes(dbus_conn, hsicache_callback, NULL);
+	}
+
+	hsi_result = hsicache_get(stream_id);
+
+	if (hsi_result == UINT32_MAX) {
```

Comment:
So, how would the end user understand that there is no such attribute because the `fwupd` service is not running?

Reply:
I added an error handling similar to the error handling for dbus connect. It sets false to every test and attaches an error message for the probe.